### PR TITLE
Minor speedup / cleanup for internal `FieldInfo` functions

### PR DESCRIPTION
**Pull request by sydney-runkle**
Simplifying some logic, and speeding up `merge_field_infos` by 14% (no major global impact, though).

### CODE REVIEW

File: pydantic/fields.py [hunk 1]
```diff
@@ -424,11 +424,6 @@ def merge_field_infos(*field_infos: FieldInfo, **overrides: Any) -> FieldInfo:
         Returns:
             FieldInfo: A merged FieldInfo instance.
         """
-        flattened_field_infos: list[FieldInfo] = []
-        for field_info in field_infos:
-            flattened_field_infos.extend(x for x in field_info.metadata if isinstance(x, FieldInfo))
-            flattened_field_infos.append(field_info)
-        field_infos = tuple(flattened_field_infos)
```
Comment (sydney-runkle) [inline pydantic/fields.py:431]: Perhaps I'm missing something, but I'm not understanding the case where we'd need this (and no tests are breaking with this change).

Comment (sydney-runkle) [reply pydantic/fields.py:431]: This still seems right to me:

```py
from pydantic import BaseModel, Field
from typing_extensions import Annotated

class Model(BaseModel):
    a: Annotated[int, Field(..., description='hi there'), Field(..., alias='A')]
    b: Annotated[str, Field(..., description='hi there'), Field(..., alias='B')] = Field(..., frozen=True)

print(Model.model_fields)
"""
{
    'a': FieldInfo(annotation=int, required=True, alias='A', alias_priority=2, description='hi there'), 
    'b': FieldInfo(annotation=str, required=True, alias='B', alias_priority=2, description='hi there', frozen=True)
}
"""
```

Comment (hramezani) [reply pydantic/fields.py:431]: hmmm, I also didn't get why we need these codes and why we don't have any breaking tests after this change.

Would be great to ask someone else for reviewing this part.

**Review by hramezani — APPROVED** [review]

File: pydantic/fields.py (hunk 1 above)
Comment (Viicos) [reply pydantic/fields.py:431]: I've tried looking at many places in the `FieldInfo` code to see if we could hit such a path, but couldn't find any for now. However, I realize the code is quite messy, and may be inconsistent in some places (e.g. the part you removed) even if things still work (there's also some code duplication). I wonder if we could simplify things here.

**Pull request merged**